# Part I: Initial Benchmark
We are going to benchmark our current implementation, first we want to set $n$ and $m$ so that they can be on a reasonably sized instance. 

We use the ks_func function we implemented in HW2, and we set $n$  and $m$  to be 1,000,000 instead of 1000 to make the memory bottleneck and performance issues more observable.  And we use the benchmark tool to see the execute time and see what'll happen.



In [19]:
using Plots
using Revise
includet("C:/Users/Odysseus/GR6104_Homework/HW3/src/ks-stat.jl")

In [20]:
using BenchmarkTools
using Random 
using Profile
using ProfileCanvas

Random.seed!(2026)
n = randn(1000000)
m = randn(1000000)
@btime ks_func(n,m)
@benchmark ks_func(n,m)


  170.601 ms (13 allocations: 91.55 MiB)


BenchmarkTools.Trial: 30 samples with 1 evaluation per sample.
 Range (min … max):  154.335 ms … 195.854 ms  ┊ GC (min … max): 0.00% … 7.74%
 Time  (median):     170.757 ms               ┊ GC (median):    4.59%
 Time  (mean ± σ):   170.900 ms ±   9.679 ms  ┊ GC (mean ± σ):  4.35% ± 3.48%

                         ▁ █     ▁                               
  ▆▁▁▁▆▁▁▆▆▁▆▆▁▆▆▁▆▆▁▆▆▆▁█▆█▁▆▆▆▆█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆▁▆▁▁▁▁▆▁▁▁▁▁▆ ▁
  154 ms           Histogram: frequency by time          196 ms <

 Memory estimate: 91.55 MiB, allocs estimate: 13.

The current implementation takes approximately 170 ms to run. More importantly, it allocates 91.55 MiB of memory. This indicates a significant inefficiency, likely due to the creation of intermediate data structures during the labeling and sorting process.